File: lib/tare.rb
require 'json'

T = JSON.parse File.read(File.expand_path('template.json', __dir__))
C = JSON.parse File.read(File.expand_path('characters.json', __dir__))

module Tare
  class << self
    private

    def render template, args
      template.gsub('{}') { args.shift }
    end

    def s outer, inner
      render T['s'], [outer, T['defaultLayout'] + inner]
    end

    def v height
      render T['s'], [T['vOuter'], T['vInner'] + T['right'] + height]
    end

    def draw commands
      commands.map do |command, *options|
        case command
        when 's'
          s T[options[0]], options[1..].map{ |x| T[x] }.join
        when 'v'
          v T[options[0]]
        end
      end
    end

    def characters
      @characters ||= C.map do |character, columns|
        [
          character,
          render(T['character'], [
            columns.count,
            character,
            columns.map{ |commands| render T['column'], [draw(commands).join] }.join,
          ]),
        ]
      end.to_h
    end
  end

  def self.html text
    render T['html'], [
      text.gsub('"', '&quot;'),
      text.upcase.chars.map{ |character| characters[character] }.compact.join,
    ]
  end
end

File: lib/tare.js
import T from './template.json';
import C from './characters.json';

const render = (template, args) => template.replaceAll('{}', () => args.shift());

const s = (outer, inner) => render(T.s, [outer, T.defaultLayout + inner]);
const v = (height) => render(T.s, [T.vOuter, T.vInner + T.right + height]);

const draw = (commands) => commands.map(([command, ...options]) => {
  switch (command) {
    case 's': return s(T[options[0]], options.slice(1).map(x => T[x]).join(''));
    case 'v': return v(T[options[0]]);
    default: throw new Error();
  }
});

const characters = new Map();
for (let [character, columns] of Object.entries(C)) {
  characters.set(character, render(T.character, [
    columns.length,
    character,
    columns.map(commands => render(T.column, [draw(commands).join('')])).join('')
  ]));
}

export default {
  html: (text) => render(T.html, [
    text.replaceAll('"', '&quot;'),
    Array.from(text.toUpperCase()).map(character => characters.get(character)).filter(x => x).join(''),
  ]),
};
